### DISCUSSION TITLE
membership questions

### DISCUSSION THREAD

**Discussion by mbj4668** [rabbitmq/ra#332] [General]
Suppose I have a cluster with 3 members, and I want to move this cluster to 3 other nodes.  I can call `add_member` with all 3 new nodes, one at the time.
Is there a way to know that it is safe to delete the 3 old members?

**Comment by kjnilsson** [comment]
ra:remove_member/2|3 does this

On Mon, 14 Nov 2022 at 07:52, Martin Bjorklund ***@***.***>
wrote:

> Suppose I have a cluster with 3 members, and I want to move this cluster
> to 3 other nodes. I can call add_member with all 3 new nodes, one at the
> time.
> Is there a way to know that it is safe to delete the 3 old members?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/ra/discussions/332>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAHFFTVWOTKZJKSU3Q5ATWIHVTDANCNFSM6AAAAAAR7QT7C4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
*Karl Nilsson*

**Comment by mbj4668** [comment]
Do you mean that remove_member returns with an error if not all state is available somewhere in the rest of the cluster?

Is there are way to test this before calling remove_member?

**Reply by michaelklishin** [reply]
A Ra cluster member can find out the most recent version of machine state by executing a consistent query. This also makes sure that a majority of nodes is online. There should be nothing else to verify besides, perhaps, any application-specific aspects of your state that Ra has no knowledge of.

**Reply by michaelklishin** [reply]
Actually, there is one more thing to verify: that the node pending removal is not the quorum-critical one. There is a [health check that does this in RabbitMQ](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/src/rabbit_quorum_queue.erl#L352-L355) that you can use example.

**Reply by mbj4668** [reply]
Ok, thanks, I'll take a closer look at that code.  From just a glance it's not obvious how to use it ;)

Just to be clear, is it always safe to call remove_node?  I.e., will it return an error if the node I  try to remove is quorum-critical?
